book/multi-arch-controller.md

```diff
@@ -0,0 +1,114 @@
+
+The reason why this works is that the `TaskRun` will create a pod that references a non-existent secret (e.g. if the `TaskRun` was called `devfile-sample-t7rd7-build-container-arm64` then the resulting pod will reference a secret called `multi-arch-ssh-devfile-sample-t7rd7-build-container-arm64`). As this secret does not exist yet, the pod will not start but instead Kubernetes will wait until the secret is ready before running the pod. This gives the controller time to provision a SSH based host, and then create the secret with the required details. The secret is created even on failure to prevent the task hanging until timeout and to allow any error messages to be propagated to the user.
+
+## The Controller
+
+The controller is responsible for allocating SSH credentials to TaskRuns that meet the contract specified above. At present, it only supports fixed pools of hosts, however work is underway to allow it to dynamically create hosts from cloud providers. It is expected that in the future it will support fixed pools, fully dynamic host provisioning (where a virtual machine is started for a single build and then destroyed), and dynamic pooling (where machines are shared, but scaled up or down on demand).
+
+### Configuration
+
+At present the controller requires no CRDs (although this might change for dynamic pooling). Configuration is provided through a `ConfigMap` that defines the available hosts. This `ConfigMap` is called `host-config` and lives in the `multi-arch-controller` namespace. This `ConfigMap` must have the `build.appstudio.redhat.com/multi-arch-config` label.
+
+An example is shown below:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    build.appstudio.redhat.com/multi-arch-config: hosts
+  name: host-config
+  namespace: multi-arch-controller
+data:
+  host1.address: "ec2-54-165-44-192.compute-1.amazonaws.com"
+  host1.secret: "awskeys"
+  host1.concurrency: "4"
+  host1.user: "ec2-user"
+  host1.arch: "arm64"
+  ppc1.address: "150.240.147.198"
+  ppc1.arch: "ppc64le"
+  ppc1.user: "root"
+  ppc1.secret: "awskeys"
+  ppc1.concurrency: "4"
+  ibmz1.address: "169.59.165.178"
+  ibmz1.arch: "s390x"
+  ibmz1.user: "root"
+  ibmz1.secret: "awskeys"
+  ibmz1.concurrency: "4"
+```
+
+Each host has a unique name, and provides different parameters that can be configured:
+
+- `address` The address or host name
+- `secret` The name of a secret in the `multi-arch-controller` namespace that contains the SSH keys for this host. These keys are never copied into the user namespace.
+- `concurrency` The number of builds that can be running on the host at any one time. Note that this is tracked by the controller rather than the host, so for example if you had a host referenced by both staging and production each environment is not aware of the other environments builds.
+- `user` The username to use to connect to the host
+- `arch` The hosts architecture
```

Comment:
In OSBS, we use the concept of platform to define the combination of a CPU architecture and an OS (just like a manifest list does). You may want to consider a similar approach here. Either as a single parameter, e.g. `platform: linux/amd64`, or two parameters, e.g. `os: linux` and `arch: amd64`.

This would also impact the "The TaskRun Contract".

Reply:
Good point.

At the moment this is just a string, so 'linux/amd64' would work fine, as long as we configured that name in the host config and pipeline runs. 

Not sure if this is worth changing the name to 'platform' rather than arch though?

Reply:
It would be much harder to change later :wink:

Reply:
@arewm What do you think? Personally I think platform makes more sense than arch, even though at the moment we only support linux.

Reply:
Platform makes sense, yes.